### PR TITLE
Add AsExpressionOrArrayComparison trait

### PR DESCRIPTION
**Pull request by hi-rustin**
Try to close https://github.com/diesel-rs/diesel/issues/2694

### CODE REVIEW

**Comment by hi-rustin** [comment]
/hold

- [ ] docs
- [ ] Resolving impl conflicts
- [ ] tests
- [ ]  check the pg manual

File: diesel/src/expression/helper_types.rs [hunk 1]
```diff
@@ -13,7 +13,7 @@ pub type SqlTypeOf<Expr> = <Expr as Expression>::SqlType;
 pub type AsExpr<Item, TargetExpr> = AsExprOf<Item, SqlTypeOf<TargetExpr>>;
 
 /// The type of `Item` when converted to an expression of `Type`
-pub type AsExprOf<Item, Type> = <Item as AsExpression<Type>>::Expression;
+pub type AsExprOf<Item, Type> = <Item as AsExpressionOrArrayComparison<Type>>::ExpressionOrArrayCmp;
```
Comment (weiznich) [inline diesel/src/expression/helper_types.rs:16]: This should be done as separate type definition. The existing `AsExprOf` impl remains useful on its own.

**Review by weiznich — COMMENTED** [review]
I would also like to see the query from the original issue added as compile fail test

**Comment by hi-rustin** [comment]
> I would also like to see the query from the original issue added as compile fail test

Of course, it's already on my to-do list.

File: diesel/src/pg/expression/array_comparison.rs [hunk 2]
```diff
@@ -94,6 +81,33 @@ where
...
+/// # Example
+///
+/// ```rust
+/// # include!("../../doctest_setup.rs");
+/// # use diesel::dsl::*;
+/// #
+/// # fn main() {
+/// #     use schema::users::dsl::*;
+/// #     let connection = establish_connection();
+/// #     connection.execute("INSERT INTO users (name) VALUES ('Jim')").unwrap();
```
Comment (TaKO8Ki) [inline diesel/src/pg/expression/array_comparison.rs:98]: These comments are not displayed in the documentation, so I think it is not necessary.

@weiznich 
Is there any particular reason why they are commented out?

**Comment by Ten0** [comment]
I have no idea why this conflicting impl error triggers considering `Any` does not implement `Expression` nor `AsExpression`.

File: diesel/src/expression/helper_types.rs (hunk 1 above)
Comment (hi-rustin) [reply diesel/src/expression/helper_types.rs]: Added a new type.

**Comment by hi-rustin** [comment]
> I have no idea why this conflicting impl error triggers considering `Any` does not implement `Expression` nor `AsExpression`.
> Also, weird, when I generate the doc for Any, this implementation shows on its page as a blanket impl:
> image
> even though no `Expression` or `AsExpression` implementation shows there.

@weiznich Do you have any thoughts on this conflict? I tried to resolve it, but I also wasn't able to find out why it conflicted.

**Comment by hi-rustin** [comment]
It looks like the GitHub actions service is out of service!

**Comment by weiznich** [comment]
@hi-rustin @Ten0 I believe that the conflicting impl error is raised because a third party crate could provide the following impl: 
```rust
impl AsExpression<crate::MyCustomThirdPartyType> for Any<crate::any_other_type> 
{
// …
}
```

From diesels point of view such a impl wouldn't be logical, but from the compiler's point of view such a impl could possibly exists.

That written, I have no clue how to workaround this.

**Comment by Ten0** [comment]
> @hi-rustin @Ten0 I believe that the conflicting impl error is raised because a third party crate could provide the following impl:
> 
> ```rust
> impl AsExpression<crate::MyCustomThirdPartyType> for Any<crate::any_other_type> 
> {
> // …
> }
> ```
> 
> From diesels point of view such a impl wouldn't be logical, but from the compiler's point of view such a impl could possibly exists.
> 
> That written, I have no clue how to workaround this.

That makes sense. One alternate way to solve this would be to introduce `.eq_any(` and `.eq_all(`.
That would be much simpler and modular from with regards to the trait-system, and I don't think there's really ever a scenario where it's going to be a blocker compared to having an `any` expression-like.
On the other hand that would break backwards-compatibility.

**Comment by weiznich** [comment]
I think going doing the extension trait variant is the only real chance to fix this issue. Method names are up to discussion, but using `.eq_any` is not possible, as the `IN` equivalent is already named that way. We should also keep in mind that this not only affects `=` but also other operators (`!= ALL()`) is also useful :wink:

**Comment by hi-rustin** [comment]
> I think going doing the extension trait variant is the only real chance to fix this issue. Method names are up to discussion, but using `.eq_any` is not possible, as the `IN` equivalent is already named that way. We should also keep in mind that this not only affects `=` but also other operators (`!= ALL()`) is also useful 😉
> Breaking things for making it more sound is fine in my opinion.

Then I understand we need to redefine the `= any()` operator for pg? Also there could be a lot of operators in this place? For example: `=` `! =` `<>` `>` `<`? Do I need to define a lot of extension traits for this to work? I'm not sure I understand you guys correctly. Thank you very much!

**Comment by hi-rustin** [comment]
I didn't have time to continue modifying it recently and closed it for now. I will open it later when I retry to fix the issue.

**Pull request closed without merging**